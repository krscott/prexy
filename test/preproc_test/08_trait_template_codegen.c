
#include <stdio.h>

#include "08_codegen.h"

int main(void)
{
    {
        char buf[80] = "Mutable string";
        struct str s = str_from_cstr(buf);

        char test;

        assert(!str_get(s, 80, &test));
        assert(!str_get(s, 15, &test));

        assert(str_get(s, 0, &test));
        assert(test == 'M');
        assert(str_get(s, 13, &test));
        assert(test == 'g');

        assert(str_set(&s, 0, 'F'));
        assert(buf[0] == 'F');
        assert(str_set(&s, 5, '!'));
        assert(buf[5] == '!');

        printf("%s\n", buf);
    }

    {
        struct strview sv = strview_from_cstr("Immutable string");

        char test;

        assert(!strview_get(sv, 80, &test));
        assert(!strview_get(sv, 16, &test));

        assert(strview_get(sv, 0, &test));
        assert(test == 'I');
        assert(strview_get(sv, 15, &test));
        assert(test == 'g');

        // strview_set() not defined

        printf("%s\n", sv.ptr);
    }

    {
        struct cstr cstr = {"Null-terminated string"};

        printf("cstr: ");
        cstr_fprint_repr(stdout, &cstr);
        printf("\n");
    }

    return 0;
}
