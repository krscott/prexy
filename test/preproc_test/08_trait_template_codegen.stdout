Futab!e string
Immutable string
